File: Week3/Insertion_Sort.cpp
#include "Sort.h"
#include <iostream>
#include <vector>
#include <fstream>
using namespace std;

class Insertion_Sort : public Sort
{
    public:
        Insertion_Sort( unsigned size ) : Sort(size) {};
        virtual void sort();

};

void Insertion_Sort::sort()
{
    for ( unsigned i = 1; i < _N; i++ )
    {
        int val = _elems[i];
        int j = i - 1;
        for ( ; j >= 0; j-- )
            if ( _elems[j] > val )
                _elems[j+1] = _elems[j];
            else
                break;
        _elems[j+1] = val;
    }
}

ostream& operator<<( ostream &os, Insertion_Sort &rhs )
{
    for ( unsigned i = 0; i < rhs.size(); i++ )
        os << rhs.elems(i) << ", ";
    return os;
}

int main()
{

    unsigned Ntest = 18;
    unsigned N = 1;
    ofstream outFile("Insertion_Sort_runtime.csv");
    outFile << "N,t" << endl;
    for ( unsigned itest = 0; itest < Ntest; itest++ )
    {

        N = N*2;
        Insertion_Sort *ss = new Insertion_Sort(N);
        ss->initialize();

        clock_t begin = clock();
        // ss->displayme();
        ss->sort();
        // ss->displayme();

        clock_t end = clock();

        cout << "N = " << ss->size() << ", RunTime: " <<  end-begin << ", isSorted?: " << ss->isSorted()  << endl;
        outFile << N << ", " << end-begin << endl;
    }
}

File: Week3/Insertion_Sort.java
import java.time.Instant;
import java.time.Duration;
public class Insertion_Sort extends Sort
{
    public Insertion_Sort ( int size )
    {
        super(size);
    }

    public void sort( )
    {
        for ( int i = 1; i < N; i++ )
        {
            int val = elems[i];
            int j = i - 1;
            for ( ; j >= 0; j-- )
            {
                if ( elems[j] > val )
                    elems[j+1] = elems[j];
                else
                    break;
            }
            elems[j+1] = val;
        }
    }

    public static void main(String[] args)
    {
        int nTest = 16;
        int N = 1;
        for ( int i = 1; i <= nTest; i++ )
        {
            N = N * 2;
            Sort ss = new Insertion_Sort(N);
            ss.initialize();

            Instant start = Instant.now();
            // ss.displayme();
            ss.sort();
            // ss.displayme();
            Instant end = Instant.now();
            Duration dur = Duration.between(start,end);

            System.out.printf("N = %d, Runtime = %d ms, isSorted? = %b\n", N, dur.toMillis(), ss.isSorted());

        }
    }
}
